README.md
# tornpaper

A Flutter Container that looks like a (maybe torn) paper

# Usage

```
TornPaper(
                key: UniqueKey(),
                backgroundColor: backgroundColor,
                hasNoise: hasNoise,
                noiseColor: noiseColor,
                hasBorder: hasBorder,
                tornWidth: tornWidth,
                tornColor: tornColor,
                tornDeepness: tornDeepness,
                seed: seedValue.toInt(),
                stepWidth: stepWidth,
                hasShadow: hasShadow,
                shadowColor: shadowColor,
                shadowOffset: Offset(shadowOffsetBottom.toDouble(), shadowOffsetRight.toDouble()),
                tornedSides: [
                  if (tornedTop) TornedSide.top,
                  if (tornedRight) TornedSide.right,
                  if (tornedLeft) TornedSide.left,
                  if (tornedBottom) TornedSide.bottom
                ],
                child: [...child...]
          )
```

Most of the variables are self explaining but some are not:

These parameters are changing the "torn"

- seed
- tornwidth
- tornDeepness
- stepWidth

The seed is the Random(seed) for the randomness.

![Torn Explanation](torn_explanation.svg?raw=true "Torn Explanation")

Try all settings on the GithubPage: https://synchronisator.github.io/tornpaper/

# TODO

- BUG: Shadow sometimes broken: https://github.com/flutter/flutter/issues/84262
- BUG: Example: Layout Overflow on small devices
- BUG: Noise unperformant -> Isolate

# Next Level:
- Tests :D
- Knüll-Optik
- Weiße Abrisskante
- "Umschlageeffekt"


CHANGELOG.md
## 0.0.2
* Documentation and liter rules 
## 0.0.1
* initial release.


lib/tornpaper.dart
library tornpaper;

import 'dart:math';

import 'package:flutter/material.dart';
import 'package:flutter/widgets.dart';
import 'package:tornpaper/clippath.dart';
import 'package:tornpaper/tornpainterbackground.dart';
import 'package:tornpaper/tornpainterborder.dart';

/// Enum with the possible Sides
enum TornedSide {
  /// Flag for a torn side left
  left,
  /// Flag for a torn side top
  top,
  /// Flag for a torn side right
  right,
  /// Flag for a torn side bottom
  bottom }

  /// TornPaper
class TornPaper extends StatefulWidget {
  /// child widget inside this Container
  final Widget child;
  /// List of sides that are torned
  final List<TornedSide> tornedSides;
  /// Randomseed for torn generation
  final int seed;
  /// step width of torn
  final int stepWidth;
  /// Color of torn
  final Color tornColor;
  /// Width of torn as meant as thickness of border
  final double tornWidth;
  /// How deep is your Torn?
  final double tornDeepness;
  /// Color of Background
  final Color backgroundColor;
  /// Has Noise or not
  final bool hasNoise;
  /// Color of Noise
  final Color noiseColor;
  /// Has Border or not
  final bool hasBorder;
  /// Has shadow or not
  final bool hasShadow;
  /// Offset of the shadow in relation to the container
  final Offset shadowOffset;
  /// Color of the shadow
  final Color shadowColor;

  /// Constructor of TornPaper
  TornPaper(
      {this.child = const SizedBox.shrink(),
      this.tornedSides = const [],
      this.seed = 369,
      this.stepWidth = 8,
      this.tornColor = Colors.black,
      this.tornWidth = 1.0,
      this.tornDeepness = 5.0,
      this.backgroundColor = Colors.white,
      this.hasNoise = false,
      this.noiseColor = Colors.black,
      this.hasBorder = true,
      this.shadowOffset = const Offset(8.0, 8.0),
      this.shadowColor = Colors.black,
      this.hasShadow = true,
      Key? key})
      : assert(stepWidth > 0), assert(tornWidth > 0), super(key: key);

  @override
  _TornPaperState createState() => _TornPaperState();
}

class _TornPaperState extends State<TornPaper> {
  // One Random for each side
  // to let the path of one side not depend on other sides
  Random _randomRight = Random(0);
  Random _randomLeft = Random(0);
  Random _randomTop = Random(0);
  Random _randomBottom = Random(0);

  //Caching
  double _lastWidth = 0.0;
  double _lastHeight = 0.0;
  Path _lastPath = Path();

  @override
  Widget build(BuildContext context) {
    return LayoutBuilder(
      builder: (context, constraints) => CustomPaint(
        foregroundPainter: TornPainterBorder(getPath(constraints),
            widget.hasBorder, widget.tornColor, widget.tornWidth),
        painter: TornPainterBackground(
            getPath(constraints),
            constraints.biggest.width,
            constraints.biggest.height,
            widget.backgroundColor,
            widget.hasNoise,
            widget.noiseColor,
            widget.hasShadow,
            widget.shadowOffset,
            widget.shadowColor),
        child: ClipPath(
            clipper: ClipPathClass(getPath(constraints)), child: widget.child),
      ),
    );
  }

  Path getPath(BoxConstraints constraints) {
    final maxWidth = constraints.biggest.width;
    final maxHeight = constraints.maxHeight;
    if (maxWidth == _lastWidth && maxHeight == _lastHeight) {
      return _lastPath;
    }

    //Reset Randoms to have stabilized Lines
    _randomRight = Random(widget.seed);
    _randomLeft = Random(widget.seed * 2);
    _randomTop = Random(widget.seed * 3);
    _randomBottom = Random(widget.seed * 4);

    final List<double> doublesRight = List.generate(maxHeight.toInt(),
        (index) => _randomRight.nextDouble() * widget.tornDeepness);
    doublesRight.first = widget.tornDeepness;
    doublesRight.last = 0.0;
    final List<double> doublesLeft = List.generate(maxHeight.toInt(),
        (index) => _randomLeft.nextDouble() * widget.tornDeepness);
    doublesLeft.first = 0.0;
    doublesLeft.last = 0.0;
    final List<double> doublesTop = List.generate(maxWidth.toInt(),
        (index) => _randomTop.nextDouble() * widget.tornDeepness);
    doublesTop.first = 0.0;
    doublesTop.last = widget.tornDeepness;
    final List<double> doublesBottom = List.generate(maxWidth.toInt(),
        (index) => _randomBottom.nextDouble() * widget.tornDeepness);
    doublesBottom.first = widget.tornDeepness;
    doublesBottom.last = widget.tornDeepness;

    final Path pathWhite = Path();
    pathWhite.moveTo(0, 0);

    // top
    if (widget.tornedSides.contains(TornedSide.top)) {
      for (var i = 0; i < doublesTop.length; i += widget.stepWidth) {
        pathWhite.lineTo(i.toDouble(), doublesTop[i]);
      }
    }
    pathWhite.lineTo(maxWidth, 0);

    // right
    if (widget.tornedSides.contains(TornedSide.right)) {
      for (var i = 0; i < doublesRight.length; i += widget.stepWidth) {
        pathWhite.lineTo(
            maxWidth - widget.tornDeepness + doublesRight[i], i.toDouble());
      }
    }
    pathWhite.lineTo(maxWidth, maxHeight);

    // bottom
    if (widget.tornedSides.contains(TornedSide.bottom)) {
      for (var i = 0; i < doublesBottom.length; i += widget.stepWidth) {
        pathWhite.lineTo(maxWidth - i.toDouble(),
            maxHeight + doublesBottom[i] - widget.tornDeepness);
      }
    }
    pathWhite.lineTo(0, maxHeight);

    //left
    if (widget.tornedSides.contains(TornedSide.left)) {
      for (var i = 0; i < doublesLeft.length; i += widget.stepWidth) {
        pathWhite.lineTo(doublesLeft[i], maxHeight - i.toDouble());
      }
    }
    pathWhite.lineTo(0, 0);
    pathWhite.close();

    _lastWidth = maxWidth;
    _lastHeight = maxHeight;
    _lastPath = pathWhite;
    return _lastPath;
  }
}


flutter_service_worker.js
'use strict';
const MANIFEST = 'flutter-app-manifest';
const TEMP = 'flutter-temp-cache';
const CACHE_NAME = 'flutter-app-cache';
const RESOURCES = {
  "version.json": "4b6db237b3514a88107a422469adfb0f",
"main.dart.js": "2e39575cf2225a55be05e3fb43f22f0f",
"index.html": "9c69948c5923ac5756a6a8a9fbd3f70d",
"/": "9c69948c5923ac5756a6a8a9fbd3f70d",
"assets/packages/cupertino_icons/assets/CupertinoIcons.ttf": "6d342eb68f170c97609e9da345464e5e",
"assets/AssetManifest.json": "2efbb41d7877d10aac9d091f58ccd7b9",
"assets/NOTICES": "40aa63e56d33dca140261e7e0173d491",
"assets/fonts/MaterialIcons-Regular.otf": "4e6447691c9509f7acdbf8a931a85ca1",
"assets/FontManifest.json": "dc3d03800ccca4601324923c0b1d6d57",
"icons/Icon-512.png": "96e752610906ba2a93c65f8abe1645f1",
"icons/Icon-192.png": "ac9a721a12bbc803b44f645561ecb1e1",
"manifest.json": "15f73b7e8a8209c2206210b3ac8dea1b",
"favicon.png": "5dcef449791fa27946b3d35ad8803796"
};

// The application shell files that are downloaded before a service worker can
// start.
const CORE = [
  "/",
"main.dart.js",
"index.html",
"assets/NOTICES",
"assets/AssetManifest.json",
"assets/FontManifest.json"];
// During install, the TEMP cache is populated with the application shell files.
self.addEventListener("install", (event) => {
  self.skipWaiting();
  return event.waitUntil(
    caches.open(TEMP).then((cache) => {
      return cache.addAll(
        CORE.map((value) => new Request(value, {'cache': 'reload'})));
    })
  );
});

// During activate, the cache is populated with the temp files downloaded in
// install. If this service worker is upgrading from one with a saved
// MANIFEST, then use this to retain unchanged resource files.
self.addEventListener("activate", function(event) {
  return event.waitUntil(async function() {
    try {
      var contentCache = await caches.open(CACHE_NAME);
      var tempCache = await caches.open(TEMP);
      var manifestCache = await caches.open(MANIFEST);
      var manifest = await manifestCache.match('manifest');
      // When there is no prior manifest, clear the entire cache.
      if (!manifest) {
        await caches.delete(CACHE_NAME);
        contentCache = await caches.open(CACHE_NAME);
        for (var request of await tempCache.keys()) {
          var response = await tempCache.match(request);
          await contentCache.put(request, response);
        }
        await caches.delete(TEMP);
        // Save the manifest to make future upgrades efficient.
        await manifestCache.put('manifest', new Response(JSON.stringify(RESOURCES)));
        return;
      }
      var oldManifest = await manifest.json();
      var origin = self.location.origin;
      for (var request of await contentCache.keys()) {
        var key = request.url.substring(origin.length + 1);
        if (key == "") {
          key = "/";
        }
        // If a resource from the old manifest is not in the new cache, or if
        // the MD5 sum has changed, delete it. Otherwise the resource is left
        // in the cache and can be reused by the new service worker.
        if (!RESOURCES[key] || RESOURCES[key] != oldManifest[key]) {
          await contentCache.delete(request);
        }
      }
      // Populate the cache with the app shell TEMP files, potentially overwriting
      // cache files preserved above.
      for (var request of await tempCache.keys()) {
        var response = await tempCache.match(request);
        await contentCache.put(request, response);
      }
      await caches.delete(TEMP);
      // Save the manifest to make future upgrades efficient.
      await manifestCache.put('manifest', new Response(JSON.stringify(RESOURCES)));
      return;
    } catch (err) {
      // On an unhandled exception the state of the cache cannot be guaranteed.
      console.error('Failed to upgrade service worker: ' + err);
      await caches.delete(CACHE_NAME);
      await caches.delete(TEMP);
      await caches.delete(MANIFEST);
    }
  }());
});

// The fetch handler redirects requests for RESOURCE files to the service
// worker cache.
self.addEventListener("fetch", (event) => {
  if (event.request.method !== 'GET') {
    return;
  }
  var origin = self.location.origin;
  var key = event.request.url.substring(origin.length + 1);
  // Redirect URLs to the index.html
  if (key.indexOf('?v=') != -1) {
    key = key.split('?v=')[0];
  }
  if (event.request.url == origin || event.request.url.startsWith(origin + '/#') || key == '') {
    key = '/';
  }
  // If the URL is not the RESOURCE list then return to signal that the
  // browser should take over.
  if (!RESOURCES[key]) {
    return;
  }
  // If the URL is the index.html, perform an online-first request.
  if (key == '/') {
    return onlineFirst(event);
  }
  event.respondWith(caches.open(CACHE_NAME)
    .then((cache) =>  {
      return cache.match(event.request).then((response) => {
        // Either respond with the cached resource, or perform a fetch and
        // lazily populate the cache.
        return response || fetch(event.request).then((response) => {
          cache.put(event.request, response.clone());
          return response;
        });
      })
    })
  );
});

self.addEventListener('message', (event) => {
  // SkipWaiting can be used to immediately activate a waiting service worker.
  // This will also require a page refresh triggered by the main worker.
  if (event.data === 'skipWaiting') {
    self.skipWaiting();
    return;
  }
  if (event.data === 'downloadOffline') {
    downloadOffline();
    return;
  }
});

// Download offline will check the RESOURCES for all files not in the cache
// and populate them.
async function downloadOffline() {
  var resources = [];
  var contentCache = await caches.open(CACHE_NAME);
  var currentContent = {};
  for (var request of await contentCache.keys()) {
    var key = request.url.substring(origin.length + 1);
    if (key == "") {
      key = "/";
    }
    currentContent[key] = true;
  }
  for (var resourceKey of Object.keys(RESOURCES)) {
    if (!currentContent[resourceKey]) {
      resources.push(resourceKey);
    }
  }
  return contentCache.addAll(resources);
}

// Attempt to download the resource online before falling back to
// the offline cache.
function onlineFirst(event) {
  return event.respondWith(
    fetch(event.request).then((response) => {
      return caches.open(CACHE_NAME).then((cache) => {
        cache.put(event.request, response.clone());
        return response;
      });
    }).catch((error) => {
      return caches.open(CACHE_NAME).then((cache) => {
        return cache.match(event.request).then((response) => {
          if (response != null) {
            return response;
          }
          throw error;
        });
      });
    })
  );
}


lib/tornpainterborder.dart
import 'package:flutter/material.dart';

/// CustomPaint for the Border
class TornPainterBorder extends CustomPainter {
  /// Outline Path
  final Path path;
  /// Has Border or not
  final bool hasBorder;
  /// Paint of the Border (color)
  late final Paint paintWhiteStroke;

  /// Constructor of the TornPainterBorder
  TornPainterBorder(
      this.path, this.hasBorder, Color colorOfTorn, double strokeWidth) {
    paintWhiteStroke = Paint()
      ..color = colorOfTorn
      ..style = PaintingStyle.stroke
      ..strokeWidth = strokeWidth;
  }

  @override
  void paint(Canvas canvas, Size size) {
    if (hasBorder) {
      canvas.drawPath(path, paintWhiteStroke);
    }
  }

  @override
  bool shouldRepaint(CustomPainter oldDelegate) =>
      (oldDelegate as TornPainterBorder).path != path;
}

/* Abrisskante in weiß

    // Paint paintPaper = Paint()
    //   ..color = Colors.black12
    //   ..style = PaintingStyle.fill
    //   ..strokeWidth = 1.0;


    // Random r = Random(123);
    // List<double> doubles = List.generate(size.height~/6, (index) => (r.nextDouble()*6)-3);

...


    // Path pathPaper = Path();
    // pathPaper.moveTo(0, 0);
    // pathPaper.lineTo(size.width-10, 0);
    // for (var i = 0; i < doubles.length; i+=stepWidth) {
    //   pathPaper.lineTo(size.width-10+doubles[i], (i*6).toDouble());
    // }
    // pathPaper.lineTo(size.width-10, size.height);
    // pathPaper.lineTo(0, size.height);
    // pathPaper.close();

...

    // canvas.drawPath(pathPaper, paintPaper);


*/


lib/tornpainterbackground.dart
import 'dart:ui';

import 'package:fast_noise/fast_noise.dart';
import 'package:flutter/material.dart';

/// CustomPaint for the Background
class TornPainterBackground extends CustomPainter {
  
  /// Path for the outline
  final Path path;
  /// width of the container to calculate the noise
  final double width;
  /// height of the container to calculate the noise
  final double height;
  /// Paint for the Background fill color
  late final Paint paintBackground;
  /// Has Noise or not
  final bool hasNoise;
  /// Color of the noise
  final Color noiseColor;
  /// Has Shadow or not
  final bool hasShadow;
  /// Color of the shadow
  final Color shadowColor;
  /// Shifted Path for Shadow, Calculated by given shadowOffset
  late final Path shiftedPath;
  /// Map with a List of Point for diffrent Colors, used to draw the noise
  late final Map<Color, List<Offset>> noisePoints = {};

  /// Constuctor of the TornPainterBackground
  TornPainterBackground(
    this.path,
    this.width,
    this.height,
    Color backgroundColor,
    this.hasNoise,
    this.noiseColor,
    this.hasShadow,
    Offset shadowOffset,
    this.shadowColor,
  ) {
    shiftedPath = path.shift(shadowOffset);
    paintBackground = Paint()
      ..color = backgroundColor
      ..style = PaintingStyle.fill;

    if (hasNoise) {
      final List<List<double>> noise = noise2(width.toInt(), height.toInt(),
          noiseType: NoiseType.Perlin,
          octaves: 5,
          frequency: 0.75,
          cellularDistanceFunction: CellularDistanceFunction.Euclidean,
          cellularReturnType: CellularReturnType.CellValue);

      for (var x = 0; x < width; x++) {
        for (var y = 0; y < height; y++) {
          // var c = (0x80 + 0x80 * noise[x][y]).floor(); // grayscale
          final offset = Offset(x.toDouble(), y.toDouble());
          if (!path.contains(offset)) {
            continue;
          }
          final color = Color.fromRGBO(
              noiseColor.red, noiseColor.green, noiseColor.blue, noise[x][y]);
          noisePoints.putIfAbsent(color, () => []);
          noisePoints[color]!.add(offset);
        }
      }
    }
  }

  @override
  void paint(Canvas canvas, Size size) {
    if (hasShadow) {
      canvas.drawShadow(shiftedPath, shadowColor, 1.0, true);
    }
    canvas.drawPath(path, paintBackground);
    if (hasNoise) {
      noisePoints.forEach((key, value) {
        canvas.drawPoints(
            PointMode.points,
            value,
            Paint()
              ..strokeWidth = 1.0
              ..color = key
              ..style = PaintingStyle.fill);
      });
    }
  }

  @override
  bool shouldRepaint(CustomPainter oldDelegate) =>
      (oldDelegate as TornPainterBackground).path != path;
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter_colorpicker/flutter_colorpicker.dart';
import 'package:tornpaper/tornpaper.dart';

void main() {
  runApp(MyApp());
}

/// Example App
class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Tornpaper Example',
      theme: ThemeData(
        primarySwatch: Colors.green,
      ),
      home: MyHomePage(title: 'Tornpaper Example'),
    );
  }
}

/// Example Mainpage
class MyHomePage extends StatefulWidget {

  /// Constructor
  MyHomePage({Key? key, required this.title}) : super(key: key);

  /// Example title
  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  //Done
  bool tornedTop = true;
  bool tornedBottom = true;
  bool tornedLeft = true;
  bool tornedRight = true;
  bool hasBorder = true;
  Color tornColor = Colors.black;
  double tornDeepness = 5.0;
  double seedValue = 200;
  int stepWidth = 8;
  double tornWidth = 1.0;
  Color backgroundColor = const Color.fromRGBO(215, 215, 160, 1.0);
  bool hasNoise = false;
  Color noiseColor = Colors.white;
  bool hasShadow = true;
  Color shadowColor = Colors.black;
  int shadowOffsetBottom = 8;
  int shadowOffsetRight = 8;

  void changeShadowColor(Color color) {
    setState(() => shadowColor = color);
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Container(
        color: Colors.green,
        child: Center(
          child: Padding(
            padding: const EdgeInsets.all(50.0),
            child: SizedBox.expand(
              child: Stack(
                children: [
                  TornPaper(
                      key: UniqueKey(),
                      backgroundColor: backgroundColor,
                      hasNoise: hasNoise,
                      noiseColor: noiseColor,
                      hasBorder: hasBorder,
                      tornWidth: tornWidth,
                      tornColor: tornColor,
                      tornDeepness: tornDeepness,
                      seed: seedValue.toInt(),
                      stepWidth: stepWidth,
                      hasShadow: hasShadow,
                      shadowColor: shadowColor,
                      shadowOffset: Offset(shadowOffsetBottom.toDouble(),
                          shadowOffsetRight.toDouble()),
                      tornedSides: [
                        if (tornedTop) TornedSide.top,
                        if (tornedRight) TornedSide.right,
                        if (tornedLeft) TornedSide.left,
                        if (tornedBottom) TornedSide.bottom
                      ],
                      child: Container()),
                  Container(
                    child: Padding(
                      padding: const EdgeInsets.all(8.0),
                      child: SingleChildScrollView(
                        child: Column(
                          mainAxisAlignment: MainAxisAlignment.center,
                          children: <Widget>[
                            const Text(
                              "TornPaper - Settings",
                              style: TextStyle(
                                  fontWeight: FontWeight.bold, fontSize: 30),
                            ),
                            Row(
                              children: const [
                                const Text("Torned Sides"),
                              ],
                            ),
                            Row(
                              children: [
                                Checkbox(
                                  value: tornedTop,
                                  onChanged: (bool? value) {
                                    setState(() {
                                      tornedTop = value!;
                                    });
                                  },
                                ),
                                const Text("Top"),
                              ],
                            ),
                            Row(
                              children: [
                                Checkbox(
                                  value: tornedBottom,
                                  onChanged: (bool? value) {
                                    setState(() {
                                      tornedBottom = value!;
                                    });
                                  },
                                ),
                                const Text("Bottom"),
                              ],
                            ),
                            Row(
                              children: [
                                Checkbox(
                                  value: tornedLeft,
                                  onChanged: (bool? value) {
                                    setState(() {
                                      tornedLeft = value!;
                                    });
                                  },
                                ),
                                const Text("Left"),
                              ],
                            ),
                            Row(
                              children: [
                                Checkbox(
                                  value: tornedRight,
                                  onChanged: (bool? value) {
                                    setState(() {
                                      tornedRight = value!;
                                    });
                                  },
                                ),
                                const Text("Right"),
                              ],
                            ),
                            Row(
                              children: [
                                const Text("Torn StepWidth"),
                                Slider(
                                  value: stepWidth.toDouble(),
                                  min: 1,
                                  max: 50,
                                  divisions: 49,
                                  label: stepWidth.round().toString(),
                                  onChanged: (double value) {
                                    setState(() {
                                      stepWidth = value.toInt();
                                    });
                                  },
                                ),
                              ],
                            ),
                            Row(
                              children: [
                                const Text("Torn Deepness"),
                                Slider(
                                  value: tornDeepness.toDouble(),
                                  min: 0,
                                  max: 20,
                                  divisions: 200,
                                  label: tornDeepness.toStringAsFixed(2),
                                  onChanged: (double value) {
                                    setState(() {
                                      tornDeepness = value;
                                    });
                                  },
                                ),
                              ],
                            ),
                            Row(
                              children: [
                                Checkbox(
                                  value: hasBorder,
                                  onChanged: (bool? value) {
                                    setState(() {
                                      hasBorder = value!;
                                    });
                                  },
                                ),
                                const Text("Has Border "),
                              ],
                            ),
                            Padding(
                              padding: const EdgeInsets.only(left: 30.0),
                              child: Row(
                                children: [
                                  const Text("Border Color "),
                                  IconButton(
                                      icon: const Icon(Icons.color_lens),
                                      color: tornColor,
                                      onPressed: !hasBorder
                                          ? null
                                          : () => showShadowColorDialog(
                                              context,
                                              tornColor,
                                              (color) => setState(
                                                  () => tornColor = color)))
                                ],
                              ),
                            ),
                            Padding(
                              padding: const EdgeInsets.only(left: 30.0),
                              child: Row(
                                children: [
                                  const Text("Torn width"),
                                  Slider(
                                    value: tornWidth.toDouble(),
                                    min: 0.1,
                                    max: 10.0,
                                    divisions: 99,
                                    label: tornWidth.round().toString(),
                                    onChanged: (double value) {
                                      setState(() {
                                        tornWidth = value;
                                      });
                                    },
                                  ),
                                ],
                              ),
                            ),
                            Row(
                              children: [
                                const Text("Seed"),
                                Slider(
                                  value: seedValue.toDouble(),
                                  min: 0,
                                  max: 1000,
                                  divisions: 500,
                                  label: seedValue.round().toString(),
                                  onChanged: (double value) {
                                    setState(() {
                                      seedValue = value;
                                    });
                                  },
                                ),
                              ],
                            ),
                            Row(
                              children: [
                                const Text("Background Color "),
                                IconButton(
                                    icon: const Icon(Icons.color_lens),
                                    color: Colors.black,
                                    onPressed: () => showShadowColorDialog(
                                        context,
                                        backgroundColor,
                                        (color) => setState(
                                            () => backgroundColor = color)))
                              ],
                            ),
                            Row(
                              children: [
                                Checkbox(
                                  value: hasShadow,
                                  onChanged: (bool? value) {
                                    setState(() {
                                      hasShadow = value!;
                                    });
                                  },
                                ),
                                const Text("Has Shadow"),
                              ],
                            ),
                            Padding(
                              padding: const EdgeInsets.only(left: 30.0),
                              child: Row(
                                children: [
                                  const Text("Shadow Color "),
                                  IconButton(
                                      icon: const Icon(Icons.color_lens),
                                      color: shadowColor,
                                      onPressed: hasShadow
                                          ? () => showShadowColorDialog(
                                              context,
                                              shadowColor,
                                              (color) => setState(
                                                  () => shadowColor = color))
                                          : null)
                                ],
                              ),
                            ),
                            Padding(
                              padding: const EdgeInsets.only(left: 30.0),
                              child: Row(
                                children: [
                                  const Text("Shadow Offset "),
                                  Column(
                                    children: [
                                      Slider(
                                        value: shadowOffsetBottom.toDouble(),
                                        min: -100,
                                        max: 100,
                                        divisions: 200,
                                        label: shadowOffsetBottom
                                            .round()
                                            .toString(),
                                        onChanged: (double value) {
                                          setState(() {
                                            shadowOffsetBottom = value.toInt();
                                          });
                                        },
                                      ),
                                      Slider(
                                        value: shadowOffsetRight.toDouble(),
                                        min: -100,
                                        max: 100,
                                        divisions: 200,
                                        label: shadowOffsetRight
                                            .round()
                                            .toString(),
                                        onChanged: (double value) {
                                          setState(() {
                                            shadowOffsetRight = value.toInt();
                                          });
                                        },
                                      ),
                                    ],
                                  ),
                                ],
                              ),
                            ),
                            const Divider(),
                            const Text("Experimental"),
                            Row(
                              children: [
                                Checkbox(
                                  value: hasNoise,
                                  onChanged: (bool? value) {
                                    setState(() {
                                      hasNoise = value!;
                                    });
                                  },
                                ),
                                const Text("Has Noise"),
                                IconButton(
                                    icon: const Icon(Icons.color_lens),
                                    color: noiseColor,
                                    onPressed: () => showShadowColorDialog(
                                        context,
                                        noiseColor,
                                        (color) => setState(
                                            () => noiseColor = color))),
                                const Text("Warning: Extremly unperformant!"),
                              ],
                            ),
                          ],
                        ),
                      ),
                    ),
                  ),
                ],
              ),
            ),
          ),
        ),
      ),
    );
  }

  void showShadowColorDialog(
      BuildContext context, Color color, Function(Color color) setColor) {
    showDialog(
      context: context,
      builder: (context) => AlertDialog(
        title: const Text('Pick a color!'),
        content: SingleChildScrollView(
          child: Column(
            children: [
              ColorPicker(
                pickerColor: color,
                onColorChanged: setColor,
                showLabel: true,
                pickerAreaHeightPercent: 0.8,
              ),
              Row(
                children: [
                  const Spacer(),
                  ElevatedButton(
                      onPressed: () {
                        setState(() {
                          Navigator.of(context).pop();
                        });
                      },
                      child: const Text("OK"))
                ],
              )
            ],
          ),
        ),
      ),
    );
  }
}
